### PR TITLE
templater: add support for evaluating method arguments

### PR DESCRIPTION
**Pull request by yuja**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (src/commands/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

**Comment by ilyagr** [comment]
Thank you, Yuya! 

If I understand correctly, the reason for this design is that you want every argument of a method to be passed as a part of the "context" to the `TemplateProperty` for the method. It took me a bit of time to figure that out; it might help to write down this idea in a comment or at least in a commit description somewhere.

I was thinking of something much simpler, but for the long term, this is certainly a better design. 👍

File: src/template_parser.rs [hunk 1]
```diff
@@ -187,29 +187,6 @@ enum Property<'a, I> {
 }
 
 impl<'a, I: 'a> Property<'a, I> {
-    fn after<C: 'a>(self, first: Box<dyn TemplateProperty<C, Output = I> + 'a>) -> Property<'a, C> {
```
Comment (martinvonz) [inline src/template_parser.rs:188]: Nice to get rid of this. I always found it ugly because it feels backwards.

File: src/template_parser.rs [hunk 2]
```diff
@@ -265,6 +265,72 @@ impl<'a, C: 'a> Expression<'a, C> {
     }
 }
 
+type OptionalArg<'i> = Option<Pair<'i, Rule>>;
+
+/// Extracts exact N required arguments.
```
Comment (martinvonz) [inline src/template_parser.rs:280]: nit: "exactly"

**Review by martinvonz — APPROVED** [review]

File: src/template_parser.rs [hunk 3]
```diff
@@ -586,16 +581,27 @@ fn parse_commit_template_rule<'a>(
...
+
+pub fn parse_commit_template<'a>(
+    repo: RepoRef<'a>,
+    workspace_id: &WorkspaceId,
+    template_text: &str,
+) -> TemplateParseResult<Box<dyn Template<Commit> + 'a>> {
+    parse_template_str(template_text, |pair| {
+        parse_commit_keyword(repo, workspace_id, pair)
+    })
+}
```
[the remark this answers is not in the document]
Comment (yuja) [reply src/template_parser.rs:700]: Yeah, that's the plan.

**Comment by ilyagr** [comment]
> Yes, and that's unavoidable because we're composing a big function of `(&Commit, formatter) -> Result<(), io::Error>` type. 

I am wondering what I can learn from this. In that spirit, the word "unavoidable" confuses me a little, since I know my crude original approach (copied below for posterity) worked. The way I understand the difference is that you really want to remove any evaluation logic from the parsing functions. That puts severe restrictions on what can go in the `formatter` part of the big function.

```diff
diff --git a/src/template_parser.rs b/src/template_parser.rs
index 7caac922..8081e94b 100644
--- a/src/template_parser.rs
+++ b/src/template_parser.rs
@@ -333,22 +333,27 @@ fn parse_boolean_method<'a>(
 
 fn parse_commit_or_change_id_method<'a>(
     name: Pair<Rule>,
-    _args: Pairs<Rule>,
+    args: Pairs<Rule>,
 ) -> Result<Property<'a, CommitOrChangeId<'a>>, TemplateParseError> {
     fn wrap_fn<'a, O>(
         f: impl Fn(&CommitOrChangeId<'a>) -> O + 'a,
     ) -> Box<dyn TemplateProperty<CommitOrChangeId<'a>, Output = O> + 'a> {
         Box::new(TemplatePropertyFn(f))
     }
+    let args = args.as_str();
     // TODO: validate arguments
+    let int_arg: Option<usize> = (!args.is_empty()).then(|| {
+        args.parse()
+            .expect("Id functions accept a single integer argument only")
+    });
     let property = match name.as_str() {
-        "short" => Property::String(wrap_fn(|id| id.short())),
-        "shortest_prefix_and_brackets" => {
-            Property::String(wrap_fn(|id| id.shortest_prefix_and_brackets()))
-        }
-        "shortest_styled_prefix" => {
-            Property::IdWithHighlightedPrefix(wrap_fn(|id| id.shortest_styled_prefix()))
-        }
+        "short" => Property::String(wrap_fn(move |id| id.short(int_arg.unwrap_or(12)))),
+        "shortest_prefix_and_brackets" => Property::String(wrap_fn(move |id| {
+            id.shortest_prefix_and_brackets(int_arg.unwrap_or(12))
+        })),
+        "shortest_styled_prefix" => Property::IdWithHighlightedPrefix(wrap_fn(move |id| {
+            id.shortest_styled_prefix(int_arg.unwrap_or(12))
+        })),
         _ => {
             return Err(TemplateParseError::no_such_method(
                 "CommitOrChangeId",
```

**Comment by yuja** [comment]
> I am wondering what I can learn from this. In that spirit, the word "unavoidable" confuses me a little, since I know my crude original approach (copied below for posterity) worked.

It works only for literals (or properties of `TemplateProperty<()>` type.)
In general, `commit_id.short(n)` is considered a function of this form `id_short(commit_id, n)`, so the argument `n` should be no different from `commit_id`.

> The way I understand the difference is that you really want to remove any evaluation logic from the parsing functions. That puts severe restrictions on what can go in the formatter part of the big function.

No. The parser doesn't know an input context, so any context-derived value can't be evaluated there. Literals could be parsed there, but I don't think such optimization is needed atm.

FWIW, I think Template/TemplateProperty needs a way to report runtime error.

**Comment by ilyagr** [comment]
This clears things up for me, thanks!

**Pull request merged**